### PR TITLE
Add support for AVX

### PR DESCRIPTION
**Pull request by bosilca**
Add support for AVX instructions for all MPI_Op.

### CODE REVIEW

**Comment by hppritcha** [comment]
The cray compile warnings/errors look legit.  Not sure why for warns and notes led to GCC returning failure however.

**Comment by bosilca** [comment]
Because my .m4 screwed up the flags. Fix on the way.

**Comment by ggouaillardet** [comment]
@bosilca I just pushed two commits to fix some typos

FWIW, I am also writing the `op/neon` component for ARM NEON (128 bits vectors).

For SVE, would you rather have it in its own module (e.g. `op/sve`) or have a single module for ARM (e.g. both NEON and SVE) ?

**Comment by bosilca** [comment]
Thanks @ggouaillardet . Regarding the test I have a way more extensive version in the pipeline, should be able to push in the next few days.

For the ARM extensions having all ARM-V* versions in a single place can be beneficial, as we will be able to reuse parts of the code. It might however make the compiling stage more complex as we will need to provide specialized flags for some of the .c file (to enable specific architecture options). What's you take on this ?

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/ibm-ompi/88b480fb98dfd2bdac939f023e1f8fe6

**Comment by ggouaillardet** [comment]
@bosilca I pushed a new fix in your PR

I added support for NEON and SVE in my own branch (based on top of this one) at https://github.com/ggouaillardet/ompi/tree/topic/op_arm
(I still need finalizing SVE flags and implement SVE runtime detection)

For now, I did two distinct components, and though there is some overlap, there is not that much imho.

If I understand correctly the `op/avx` component, and let's say we run on the latest xeon and have to `MPI_SUM` n+63 `int8_t`, that would be
 - 3*n AVX512 instructions (n iterations), this is the main part
- 3 AVX256 instructions (1 iteration)
- 3 SSE (AVX128?) instructions (1 iteration)
- 15*3 scalar instructions (2 iterations, 8+7 unrolling)

before each step, the flags are tested, and draining the main loop (e.g. the last 63 `int8_t`) will invoke two loops with 0 or 1 iterations, which is quite some overhead that could be avoided. Also, and unless this has changed, I remember Intel discouraged mixing AVX and SSE instructions in the same program.
my gut feeling is this could be faster if
 - we use one subroutine per avx type
 - flags are only tested once, and functions pointers are set accordingly
 - draining the loop is done in a single scalar loop

we could also do loop peeling to be faster when data is not optimally aligned.

In theory, we could have several components (e.g. one per avx type) but in order to avoid code duplication, it can be left in a single module (and since subroutine selection is performed at selection time, there is no runtime overhead)

**Comment by jjhursey** [comment]
bot:ibm:xl:retest

**Comment by bosilca** [comment]
@ggouaillardet I did not hear about anything but performance issues while mixing AVX and SSE, and these are not a concern for us because in the worst case we do such a transition once. We could follow [Intel's suggestions](https://software.intel.com/en-us/articles/avoiding-avx-sse-transition-penalties) and pass the code through vtune to make sure.

The code can certainly be improved, but I don't know by how much. I assumed that most of the MPI_Op executes on data large enough to completely hide the few cycles lost in the branch mismanagement, so that reordering and loop peeling might not be the such a deal breaker. But we might want to investigate how to deal with the case where the user data alignment matches the AVX needs, as this should allow us to use AVX instructions for aligned memory (_load_ instead of _loadu_) accesses but I don't how much we can gain. Some hits can be found [here](https://software.intel.com/en-us/forums/intel-isa-extensions/topic/752392) on McCalpin answer.

For the SVE code, @dong0321 and myself were working on a branch in my repo, that you can access via a [PR](https://github.com/bosilca/ompi/pull/14). It is mostly based on the AVX code, and has no support for Neon. We should merge our efforts on this, and reach out to Fujitsu  (@Shinji-Sumimoto) and ARM (@shamisp) for final validation.

**Comment by bosilca** [comment]
I don't understand the AZR issue here. It seems that the instruction is not supported, but it did find it while looking for it during configure. So I wonder of somehow the additional flags detected during configure are not passed to the compiler for the avx_functions.c file. But for this I would need to see the compile arguments used. Can someone form Mellanox send me this info please.

**Comment by shamisp** [comment]
@bosilca you can use ```armie``` on any arm to validate correctness. You can run it on any armv8. @dong0321 is master of it :)

**Comment by ggouaillardet** [comment]
@bosilca the AZR issue is legit.

long story short, `_mm256_max_epi32()` is an AVX2 instruction (and is only used when available), but `_mm256_max_epi64()` is an AVX512 instruction (and is used when AVX2 is available, even if AVX512 is not)

we should definitely merge our efforts with respect to SVE (and fwiw, we ended up with very similar code)

last time I checked about alignment impact on x86_64 (sandy bridge ?), I found there is no point into using aligned instructions (read an unaligned instruction is just as fast as the aligned instruction when the data is aligned; and when the data is not aligned, the former is slower and the latter crashes) and I also noticed the Intel compilers did not even bother generating aligned instructions even if pragmas were used to tell the data is aligned.

**Comment by bosilca** [comment]
I rebase to reduce the number of commits. There are few things that still need to be fixed:

- [ ] on some architecture the code generates illegal instructions. I don't know yet why, but we try to execute vmovdqa64 on a machine without AVX512.
- [ ] Support for unsigned int if incomplete.

**Comment by ggouaillardet** [comment]
@bosilca is the AVX512 instruction generated by an AVX512 intrinsic or some plain C code? If the latter we might have no choice but having one `op_avx_functions` file per AVX flavor (unless there is a compiler flag to tell the compiler only intrinsics can generate AVX512 instructions)

**Comment by bosilca** [comment]
@ggouaillardet I think I finally got it right. Here is the logic:
- we support 3 x86 ISA: avx512f + wb, AVX2, and AVX (+sse4.1+sse3). We detect if we are generating code for x86 and if the compiler can generate these 3 flavors in configure.
- we recompile the code 3 times, each time with a different flag and a set of defines to reflect what we discovered during configure.
- we merge all these files into a single library
- at runtime we pick for each MPI_OP the best flavor supported by the current CPU (in order avx512 > avx2 > avx)

**Comment by awlauria** [comment]
bot:ompi:retest

**Comment by jsquyres** [comment]
bot:ompi:retest

**Comment by jsquyres** [comment]
@bosilca @ggouaillardet Is this ready to be merged?  The goal is to make a v4.1.x RC next Monday (and Friday is a holiday for most in the US).

**Comment by bosilca** [comment]
It seems to compile everywhere, and as far as I can see it works. @dong0321 will give it a more intensive testing over the weekend.

**Comment by jsquyres** [comment]
bot:aws:retest

**Comment by jsquyres** [comment]
@bosilca In the `--disable-dlopen` AWS CI build, we're getting this error:

```
  CC       ompi_info.o
  CC       param.o
  GENERATE ompi_info.1
  CCLD     ompi_info
../../../ompi/.libs/libmpi.so: undefined reference to `ompi_op_avx_functions_avx512'
../../../ompi/.libs/libmpi.so: undefined reference to `ompi_op_avx_3buff_functions_avx512'
```

**Comment by bosilca** [comment]
I need a little bit more info, I can't replicate. It is strange that only the 512 part is missing, but the AVX2 is in.

**Comment by jsquyres** [comment]
https://jenkins.open-mpi.org/jenkins/job/open-mpi.build.configure_options/6739/CONFIGURE_OPTIONS=--disable-dlopen/consoleFull

**Comment by ggouaillardet** [comment]
That was a `#ifdef` vs `#if` issue and I just pushed a fix for that.
Feel free to squash before merge.

**Comment by jsquyres** [comment]
bot:aws:retest

Same real-but-unrelated-to-this-PR timeout we've been seeing for a while (see #7847).

**Comment by jsquyres** [comment]
bot:aws:retest

**Comment by jsquyres** [comment]
# Intel icc 19.04

When I compile with icc 19.04, I get the following (which is probably unsurprising):

```
# After fixing the `configure.m4` errors noted in the PR review
3274 --- MCA component op:avx (m4 configuration macro)
3275 checking for MCA component op:avx compile mode... dso
3276 checking for AVX512 support (no additional flags)... yes
3277 checking for AVX2 support (no additional flags)... yes
3278 checking for AVX support (no additional flags)... yes
3279 checking for SSE4.1 support... yes
3280 checking for SSE3 support... yes
3281 checking if MCA component op:avx can compile... yes
```

But then I get this when compiling nearly every file in Open MPI:

```
  CC       opal_datatype_resize.lo
In file included from ../../../opal/include/opal/sys/atomic.h(64),
                 from ../../../opal/mca/threads/thread_usage.h(32),
                 from ../../../opal/class/opal_object.h(126),
                 from ../../../opal/datatype/opal_datatype.h(44),
                 from ../../../opal/datatype/opal_convertor.h(35),
                 from ../../../opal/datatype/opal_convertor_internal.h(21),
                 from ../../../opal/datatype/opal_datatype_pack.c(28):
../../../opal/include/opal/sys/atomic_stdc.h(119): warning #2330: argument of type "opal_a\
tomic_int32_t={_Atomic(int32_t={int})} *" is incompatible with parameter of type "volatile\
 void *" (dropping qualifiers)
  OPAL_ATOMIC_STDC_DEFINE_FETCH_OP(add, 32, int32_t, +)
  ^
```

These are just warnings, but it makes the output from `make` be 60+ MB of this kind of output.

# GNU Gcc 7.3.0, 8.2.0, and 9.3.0

With all 3 of these versions of GCC, I get the following:

```
# After fixing configure.m4 as noted in the review:
3287 --- MCA component op:avx (m4 configuration macro)
3288 checking for MCA component op:avx compile mode... dso
3289 checking for AVX512 support (no additional flags)... no
3290 checking for AVX512 support (with -march=skylake-avx512)... yes
3291 checking for AVX2 support (no additional flags)... no
3292 checking for AVX2 support (with -mavx2)... yes
3293 checking for AVX support (no additional flags)... yes
3294 checking for SSE4.1 support... no
3295 checking for SSE3 support... no
3296 checking for AVX support (with -mavx)... yes
3297 checking for SSE4.1 support... yes
3298 checking for SSE3 support... yes
3299 checking if MCA component op:avx can compile... yes
```

And then:

```
$ cd ompi/op/mca/avx
$ make
1    CC       op_avx_component.lo
2    CC       liblocal_ops_avx_la-op_avx_functions.lo
3    CCLD     liblocal_ops_avx.la
4    CC       liblocal_ops_avx2_la-op_avx_functions.lo
/tmp/ccN7HHy9.s: Assembler messages:
/tmp/ccN7HHy9.s:1583: Error: no such instruction: `vinserti128 $0x1,16(%rdi,%rax),%ymm3,%ymm1'
/tmp/ccN7HHy9.s:1584: Error: no such instruction: `vinserti128 $0x1,16(%rsi,%rax),%ymm2,%ymm0'
/tmp/ccN7HHy9.s:1585: Error: suffix or operands invalid for `vpaddd'
/tmp/ccN7HHy9.s:1587: Error: no such instruction: `vextracti128 $0x1,%ymm0,16(%rsi,%rax)'
/tmp/ccN7HHy9.s:1730: Error: no such instruction: `vinserti128 $0x1,16(%rdi,%rax),%ymm3,%ymm1'
/tmp/ccN7HHy9.s:1731: Error: no such instruction: `vinserti128 $0x1,16(%rsi,%rax),%ymm2,%ymm0'
/tmp/ccN7HHy9.s:1732: Error: suffix or operands invalid for `vpaddd'
/tmp/ccN7HHy9.s:1734: Error: no such instruction: `vextracti128 $0x1,%ymm0,16(%rsi,%rax)'
/tmp/ccN7HHy9.s:1877: Error: no such instruction: `vinserti128 $0x1,16(%rdi,%rax),%ymm3,%ymm0'
/tmp/ccN7HHy9.s:1878: Error: no such instruction: `vinserti128 $0x1,16(%rsi,%rax),%ymm2,%ymm1'
...etc.
```

This is on an older Sandy Bridge architecture machine, so it could be before AVX2 was available...?

My full configure output from gcc 9.3.0 and corresponding config.log is attached.

[configure-output.txt](https://github.com/open-mpi/ompi/files/4873754/configure-output.txt)
[config.log](https://github.com/open-mpi/ompi/files/4873755/config.log)

File: ompi/mca/op/avx/configure.m4 [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                            [[
+    int A[4] = {0, 1, 2, 3};
+    __m128i vA = _mm_lddqu_si128((__m128i*)&A)
+                            ]])],
+                    [op_sse3_support=1
+                     AC_MSG_RESULT([yes])],
+                    [AC_MSG_RESULT([no])])
+                ])
+           # Second pass, do we need to add the AVX flag ?
+           AS_IF([test $op_avx_support -eq 0 -o $op_sse41_support -eq 0 -o $op_sse3_support -eq 0],
```
[the remark this answers is not in the document]
Comment (bosilca) [reply ompi/mca/op/avx/configure.m4:141]: We try to find if there is a need for look for a special flag, so we need to check if any of them failed at the prior stage. -o seems totally legit to me, why is there a need for '||' instead ?

File: ompi/mca/op/avx/configure.m4 [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+                AS_IF([test $op_sse41_support -eq 0],
+                    [AC_MSG_CHECKING([for SSE4.1 support])
+                     AC_LINK_IFELSE(
+                         [AC_LANG_PROGRAM([[#include <immintrin.h>]],
+                                 [[
+    __m128i vA, vB;
+    (void)_mm_max_epi8(vA, vB)
+                                 ]])],
+                         [op_sse41_support=1
+                          AC_MSG_RESULT([yes])],
```
[the remark this answers is not in the document]
Comment (bosilca) [reply ompi/mca/op/avx/configure.m4:167]: That's already done few lines above. There is a comment at line 92 that explains the logic. Basically, here we know we need the -mavx flag, we check to see if it gives us access to the SSE3 and SSE4.1 instructions.

File: ompi/mca/op/avx/configure.m4 [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+                AS_IF([test $op_sse3_support -eq 0],
+                    [AC_MSG_CHECKING([for SSE3 support])
+                     AC_LINK_IFELSE(
+                         [AC_LANG_PROGRAM([[#include <immintrin.h>]],
+                                 [[
+    int A[4] = {0, 1, 2, 3};
+    __m128i vA = _mm_lddqu_si128((__m128i*)&A)
+                                 ]])],
+                         [op_sse3_support=1
+                          AC_MSG_RESULT([yes])],
```
[the remark this answers is not in the document]
Comment (bosilca) [reply ompi/mca/op/avx/configure.m4:179]: same answer as above.

File: ompi/mca/op/avx/op_avx_component.c [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+}
+
+/*
+ * Register MCA params.
+ */
+static int
+avx_component_register(void)
+{
+    int32_t requested_flags;
+    requested_flags = mca_op_avx_component.flags = has_intel_AVX_features();
```
[the remark this answers is not in the document]
Comment (bosilca) [reply ompi/mca/op/avx/op_avx_component.c:178]: For safety mostly. Let me change to make this clear.

File: ompi/mca/op/avx/op_avx_component.c [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    (void) mca_base_component_var_register(&mca_op_avx_component.super.opc_version,
+                                           "support",
+                                           "Level of SSE/MMX/AVX support to be used (combination of processor capabilities as follow SSE 0x01, SSE2 0x02, SSE3 0x04, SSE4.1 0x08, AVX 0x010, AVX2 0x020, AVX512F 0x100, AVX512BW 0x200) capped by the local architecture capabilities",
+                                           MCA_BASE_VAR_TYPE_INT, NULL, 0, 0,
+                                           OPAL_INFO_LVL_9,
```
[the remark this answers is not in the document]
Comment (bosilca) [reply ompi/mca/op/avx/op_avx_component.c:183]: I don't think anybody should touch these. I added the MCA mostly for me, so that I can test all cases. But they could be tuned by a tool, so I'll bump them up to 6.

File: ompi/mca/op/avx/configure.m4 (hunk 1 above)
[the remark this answers is not in the document]
Comment (jsquyres) [reply ompi/mca/op/avx/configure.m4:141]: OR vs. AND: gotcha.

|| vs -o: Using `||` is more portable.  I have recollections of some flavors of test (maybe on BSDs?) that do not support `-o` or `-a`.

File: ompi/mca/op/avx/configure.m4 (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsquyres) [reply ompi/mca/op/avx/configure.m4:167]: I mis-parsed the code.  I grok it now.

File: test/datatype/check_op.sh [hunk 6]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash
+
```
[the remark this answers is not in the document]
Comment (bosilca) [reply test/datatype/check_op.sh:2]: I dont want this test to fail on the first failure. I want to see all failures.

File: ompi/mca/op/avx/configure.m4 [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    __m128 vA, vB;
+    _mm_add_ps(vA, vB)
+                       ]])],
+               [op_avx_support=1
+                AC_MSG_RESULT([yes])],
+               [AC_MSG_RESULT([no])])
+           #
+           # Check for SSE4.1 support
+           #
+           AS_IF([$op_avx_support -eq 1],
```
[the remark this answers is not in the document]
Comment (bosilca) [reply ompi/mca/op/avx/configure.m4:113]: good catch.

**Comment by bosilca** [comment]
The issue reported by the Intel compiler has nothing to do with this patch. It's about the compiler being vocal about a volatile being passed as an Atomic.

**Comment by bosilca** [comment]
for the second case can you please check the flags used to compile the AVX2 library. All these instructions are part of the AVX2 support, it looks as if the -mavx2 flag was not passed to the compiler.

File: ompi/mca/op/avx/configure.m4 (hunk 1 above)
[the remark this answers is not in the document]
Comment (bosilca) [reply ompi/mca/op/avx/configure.m4:141]: OK, I'll change to ||

**Comment by jsquyres** [comment]
Looks like `-mavx2` is there (left it wrapped in the display below for readability) -- this is the output from `make V=1`:

```
13 libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../../../../../ompi/mca/op/avx -I../../../\
   ../opal/include -I../../../../ompi/include -I../../../../oshmem/include -I../../../../o\
   pal/mca/hwloc/hwloc2/hwloc/include/private/autogen -I../../../../opal/mca/hwloc/hwloc2/\
   hwloc/include/hwloc/autogen -I../../../../ompi/mpiext/cuda/c -iquote../../../../.. -iqu\
   ote../../../.. -iquote../../../../../opal/include -iquote../../../../../ompi/include -i\
   quote../../../../../oshmem/include -I/home/jsquyres/git/ompi/build-gcc930/opal/mca/pmix\
   /pmix4x/openpmix/include -I/home/jsquyres/git/ompi/opal/mca/pmix/pmix4x/openpmix/includ\
   e -I/home/jsquyres/git/ompi/build-gcc930/opal/mca/event/libevent2022/libevent/include -\
   I/home/jsquyres/git/ompi/opal/mca/event/libevent2022/libevent -I/home/jsquyres/git/ompi\
   /opal/mca/event/libevent2022/libevent/include -I/home/jsquyres/git/ompi/build-gcc930/op\
   al/mca/hwloc/hwloc2/hwloc/include -I/home/jsquyres/git/ompi/opal/mca/hwloc/hwloc2/hwloc\
   /include -I/usr/local/include -I/usr/local/include -mavx2 -DGENERATE_SSE3_CODE -DGENERA\
   TE_SSE41_CODE -DGENERATE_AVX_CODE -DGENERATE_AVX2_CODE -O3 -DNDEBUG -finline-functions \
   -fno-strict-aliasing -mcx16 -MT liblocal_ops_avx2_la-op_avx_functions.lo -MD -MP -MF .d\
   eps/liblocal_ops_avx2_la-op_avx_functions.Tpo -c ../../../../../ompi/mca/op/avx/op_avx_\
   functions.c  -fPIC -DPIC -o .libs/liblocal_ops_avx2_la-op_avx_functions.o
14 /tmp/ccmZoSSt.s: Assembler messages:
15 /tmp/ccmZoSSt.s:1593: Error: no such instruction: `vinserti128 $0x1,-16(%rdi),%ymm0,%ymm1'
16 /tmp/ccmZoSSt.s:1595: Error: no such instruction: `vinserti128 $0x1,-16(%rax),%ymm0,%ymm0'
```

**Comment by jsquyres** [comment]
Files #7909 about the intel compiler atomic warnings.

File: test/datatype/check_op.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (jsquyres) [reply test/datatype/check_op.sh:2]: The `make check` framework will keep running all the rest of the tests in this directory, even if one of them fails.

If you don't put something like `set -euo pipefail` in your script, then you need to check the exit status of (nearly?) every command, which you aren't doing.  Otherwise, the test will fail inside this script, but then `make check` won't recognize it as failed, and the final count of past/fail/skip at the end of this directory will be inaccurate (i.e., silent test fails).

All of these options are so useful in catching errors that I generally `set -euo pipefail` in all my bash scripts these days...

File: ompi/mca/op/avx/configure.m4 [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+                AC_LINK_IFELSE(
+                    [AC_LANG_PROGRAM([[#include <immintrin.h>]],
+                            [[
+    __m128 vA, vB;
+    _mm_add_ps(vA, vB)
+                            ]])],
+                    [op_avx_support=1
+                     MCA_BUILD_OP_AVX_FLAGS="-mavx"
+                     op_sse41_support = 0
+                     op_sse3_support = 0
```
[the remark this answers is not in the document]
Comment (jsquyres) [reply ompi/mca/op/avx/configure.m4:154]: Note: these two configure.m4 assignment statements are still not fixed.

This is not a style issue -- it's a correctness issue: `a = b` (with spaces) is not a valid Bourne shell assignment:

```
$ cat foo.sh
#!/bin/sh
a=1
a = 2
echo $a

$ ./foo.sh
./foo.sh: line 4: a: command not found
1
```

**Comment by jsquyres** [comment]
@bosilca With a bunch of trial and error, this program fails to compile for me on my systems:

```
$ cat avx.c
#include <immintrin.h>

int main()
{
    void *in1;
    __m256i vA = _mm256_loadu_si256((__m256i*) in1);

    return 0;
}

$ gcc -mavx2 avx.c -o avx
/tmp/ccLxzMsm.s: Assembler messages:
/tmp/ccLxzMsm.s:18: Error: no such instruction: `vinserti128 $0x1,16(%rax),%ymm0,%ymm0'
```

Calling functions like `_mm256_add_epi32()` seem to be fine -- it's the `_mm256_loadu_si256()` that appears to be the problem.

**Comment by bosilca** [comment]
We figured it out, Jeff's assembler (v2.20.51.0.2) was from an older gcc version. With a newer version everything worked as expected. I will add a test to make sure I catch such corner cases, and disable AVX2 support to allow the build process to reach completion.

File: test/datatype/check_op.sh (hunk 6 above)
[the remark this answers is not in the document]
Comment (bosilca) [reply test/datatype/check_op.sh:2]: We did not use this script in `make check`. I pipe the result through grep to see how many tests failed, as the script test all MPI_Op for all types of predefined datatypes.

[the remark this answers is not in the document]
Comment (jsquyres) [reply test/datatype/check_op.sh:2]: Ah, I see.

You still might want to `set -u` to catch unbound variables.  Very helpful at preventing/finding bugs.

**Comment by jsquyres** [comment]
Just to clarify:

* I was testing on RHEL 6
* I had manually installed (i.e., compiled from source) several versions of GCC:
    * GCC 7.3.0
    * GCC 8.2.0
    * GCC 9.3.0
* The assembler (`as`) is not part of GCC; it is part of the `binutils` package.  So all of these compilers were using the `as` that shipped with RHEL 6:
```
$ as --version
GNU assembler version 2.20.51.0.2-5.36.el6 20100205
Copyright 2009 Free Software Foundation, Inc.
...
```

Case in point:
* If I build/install the newest binutils (2.34), and
* Re-install GCC 9.3.0 from source -- forcing it to use the `as` v2.34
* Then everything works fine.
* Indeed, if you look closely at the error message, it's the assembler that is complaining about a non-existent instruction.  That's because the RHEL 6-default `as` is ancient (from 2009), and simply does not recognize the AVX2 instruction.  Upgrading to a newer `as` solves the problem.

The fix is to strengthen OMPI's avx `configure.m4` test to make sure that the problematic instruction can, indeed, compile and link.

**Comment by jsquyres** [comment]
@bosilca Can you apply the same kind of fix for the 512-sized vectors that you did for the 256-sized vectors?  I'm now getting errors with ``_mm512_loadu_si512()` similar to what we saw yesterday:

```
$ cat avx.c
#include <immintrin.h>

int main()
{
    int in1[9] = {0, 1, 2, 3, 4, 5, 6, 7, 8};
    __m512i vA = _mm512_loadu_si512((__m256i*) &(in1[1]));

    return 0;
}

$ gcc -march=skylake-avx512 avx.c -o avx
/tmp/ccYemRMl.s: Assembler messages:
/tmp/ccYemRMl.s:28: Error: no such instruction: `vmovdqu64 (%rax),%zmm0'
/tmp/ccYemRMl.s:29: Error: no such instruction: `vmovdqa64 %zmm0,-56(%rsp)'
```

**Comment by jsquyres** [comment]
More bad news, I'm afraid.  ☹️ 

I ran the reduce_local test with gcc 9+newest `as` (i.e., all the AVX support *did* compile on my platform) on my Sandy Bridge platform, and got near-complete failures when AVX was actually used.  Here's one result from me running manually:

```
$ mpirun -np 1 reduce_local 1048583 i 16 min
MPI_SUM MPI_UINT8_T 8  count  1  time 0.000005 seconds
MPI_SUM MPI_UINT8_T 8  count  2  time 0.000000 seconds
MPI_SUM MPI_UINT8_T 8  count  4  time 0.000000 seconds
MPI_SUM MPI_UINT8_T 8  count  8  time 0.000000 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  16  time 0.000000 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  32  time 0.000000 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  64  time 0.000000 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  128  time 0.000000 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  256  time 0.000000 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  512  time 0.000000 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  1024  time 0.000000 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  2048  time 0.000000 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  4096  time 0.000000 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  8192  time 0.000001 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  16384  time 0.000001 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  32768  time 0.000003 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  65536  time 0.000005 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  131072  time 0.000012 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  262144  time 0.000025 seconds
First error at position 0 (5 sum 253 != 255)
MPI_SUM MPI_UINT8_T [fail] count  524288  time 0.000066 seconds
```

Here's the stdout from configure in this environment:

```
3291 --- MCA component op:avx (m4 configuration macro)
3292 checking for MCA component op:avx compile mode... dso
3293 checking for AVX512 support (no additional flags)... no
3294 checking for AVX512 support (with -march=skylake-avx512)... yes
3295 checking for AVX2 support (no additional flags)... no
3296 checking for AVX2 support (with -mavx2)... yes
3297 checking if _mm256_loadu_si256 generates code that can be compiled... yes
3298 checking for AVX support (no additional flags)... yes
3299 checking for SSE4.1 support... no
3300 checking for SSE3 support... no
3301 checking for AVX support (with -mavx)... yes
3302 checking for SSE4.1 support... yes
3303 checking for SSE3 support... yes
3304 checking if MCA component op:avx can compile... yes
```

And you can see the propagation of those flags in a rendered Makefile:

```
628 MCA_BUILD_OP_AVX2_FLAGS = -mavx2
629 MCA_BUILD_OP_AVX512_FLAGS = -march=skylake-avx512
630 MCA_BUILD_OP_AVX_FLAGS = -mavx
```

Is building with `-march=skylake-avx512` advisable on an older platform (i.e., Sandy Bridge)?

**Comment by rhc54** [comment]
Why don't you just detect the platform and turn it "off" for anything older than skylake?? It isn't worth all the trouble for these corner cases.

**Comment by bosilca** [comment]
I will modify the 512 case as well, not because I think it's the right approach, but because I want to get this done. While on the 256 case I understand that `-mavx2` might not be enough to cover all the instructions in the generated code, for the 512 case I am using `-march=`, and this should generate __exactly__ the same code everywhere. So, we can keep adding bandaid to OMPI as we want, but real issue is still there.

**Comment by bosilca** [comment]
@jsquyres the test is correct. As I mentionned the AVX instructions use a different overflow policy, the resulting value is set to the min/max of the representative value for the type. Thus, unsigned 8 bits 5 + 253 will be 2, except with AVX where the result will be 255. Good thing that MPI does not mandate how these operations behave with regard to overflow. I will change the test to avoid printing this information.

**Comment by jsquyres** [comment]
Sorry for the delay; I finally got to testing this PR again this morning.  It looks good!

I pushed a commit to this PR that fixes up the check_ops.sh script -- feel free to squash if you like it.

I found that on my Sandy Bridge machines:

- gcc 9 + new `as`: compiles fine
- gcc 8 + old `as`: compiles fine
- icc 19: compiles fine

Yay!

But after I fixed up the check_ops script, I noticed that all three cases are failing MIN and MAX with 64-bit values:

```
$ ./check_op.sh |& tee out.txt
...lots of output...
$ $ grep fail out.txt
MPI_MAX    MPI_UINT64_T [fail] count  1048576     time 0.003099 seconds
MPI_MAX    MPI_UINT64_T [fail] count  1048577     time 0.002463 seconds
MPI_MAX    MPI_UINT64_T [fail] count  1048583     time 0.002330 seconds
MPI_MAX    MPI_UINT64_T [fail] count  1048591     time 0.002797 seconds
MPI_MAX    MPI_UINT64_T [fail] count  1048607     time 0.003085 seconds
MPI_MAX    MPI_UINT64_T [fail] count  1048639     time 0.002938 seconds
MPI_MAX    MPI_UINT64_T [fail] count  1048703     time 0.002454 seconds
MPI_MAX    MPI_UINT64_T [fail] count  1048706     time 0.002121 seconds
MPI_MIN    MPI_UINT64_T [fail] count  1048576     time 0.001572 seconds
MPI_MIN    MPI_UINT64_T [fail] count  1048577     time 0.001552 seconds
MPI_MIN    MPI_UINT64_T [fail] count  1048583     time 0.001639 seconds
MPI_MIN    MPI_UINT64_T [fail] count  1048591     time 0.002461 seconds
MPI_MIN    MPI_UINT64_T [fail] count  1048607     time 0.001633 seconds
MPI_MIN    MPI_UINT64_T [fail] count  1048639     time 0.001597 seconds
MPI_MIN    MPI_UINT64_T [fail] count  1048703     time 0.001584 seconds
MPI_MIN    MPI_UINT64_T [fail] count  1048706     time 0.001635 seconds
```

**Comment by bosilca** [comment]
A typo in the tester (max and min were not correctly computed). I merged everything into a single commit, this is ready to go.

**Comment by jsquyres** [comment]
bot:aws:retest

**Comment by jsquyres** [comment]
🎉
